File: vh_manual_teleop/vh_network/src/video_encoder.py
#!/usr/bin/env python3
import rclpy
from rclpy.node import Node
from rclpy.qos import QoSProfile, ReliabilityPolicy, HistoryPolicy
from sensor_msgs.msg import Image
from cv_bridge import CvBridge
import cv2
import time

class VideoEncoder(Node):
    def __init__(self):
        super().__init__('video_encoder')
        self.bridge = CvBridge()
        self.out = None
        
        self.rs_ip = '10.0.0.2' 
        self.port = 5006
        self.fps = 30
        
        # Define o perfil QoS para coincidir com o do AWSIM
        qos_profile = QoSProfile(
            reliability=ReliabilityPolicy.BEST_EFFORT,
            history=HistoryPolicy.KEEP_LAST,
            depth=10
        )

        # Aplica o perfil à subscrição
        self.sub = self.create_subscription(
            Image, 
            '/sensing/camera/traffic_light/image_raw', 
            self.encode_callback, 
            qos_profile)
        
        
        self.get_logger().info(f"Nó Encoder iniciado. A enviar para {self.rs_ip}:{self.port}")

    def encode_callback(self, msg):
        start_time = time.time()
        
        # 1. Converter imagem
        cv_image = self.bridge.imgmsg_to_cv2(msg, desired_encoding='bgr8')
        height, width, _ = cv_image.shape

        # 2. Inicializar o pipeline GStreamer (apenas no primeiro frame)
        if self.out is None:
            pipeline = (
                f"appsrc ! videoconvert ! video/x-raw,format=I420 ! "
                f"nvh264enc preset=low-latency rc-mode=cbr bitrate=5000 zerolatency=true ! "
                f"rtph264pay config-interval=1 pt=96 ! "
                f"udpsink host={self.rs_ip} port={self.port} sync=false"
            )
            self.out = cv2.VideoWriter(pipeline, cv2.CAP_GSTREAMER, 0, self.fps, (width, height), True)
            if not self.out.isOpened():
                self.get_logger().error("Falha ao abrir o pipeline GStreamer.")
                return

        # 3. Adicionar o Timestamp visual para cálculo de latência end-to-end
        current_time_str = f"Time: {time.time():.3f}"
        cv2.putText(cv_image, current_time_str, (10, 30), cv2.FONT_HERSHEY_SIMPLEX, 1, (0, 255, 0), 2)

        # 4. Enviar para a GPU e depois UDP
        self.out.write(cv_image)
        
        # 5. Calcular tempo de processamento local
        process_time_ms = (time.time() - start_time) * 1000
        self.get_logger().info(f"Latência de codificação: {process_time_ms:.1f} ms") 

    def destroy_node(self):
        if self.out is not None:
            self.out.release()
        super().destroy_node()

def main(args=None):
    rclpy.init(args=args)
    node = VideoEncoder()
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
    finally:
        node.destroy_node()
        rclpy.shutdown()

if __name__ == '__main__':
    main()

File: vh_camera/src/video_encoder.cpp
#include <rclcpp/rclcpp.hpp>
#include <sensor_msgs/msg/image.hpp>
#include <cv_bridge/cv_bridge.h>
#include <opencv2/opencv.hpp>
#include <string>

class VideoEncoderTX : public rclcpp::Node {
public:
    VideoEncoderTX() : Node("video_encoder_tx_cpp"), writer_initialized_(false) {
        this->declare_parameter<std::string>("ip_address", "10.0.0.1");
        this->declare_parameter<int>("port", 5007);
        this->declare_parameter<int>("bitrate", 5000); 
        
        ip_address_ = this->get_parameter("ip_address").as_string();
        port_ = this->get_parameter("port").as_int();        
        bitrate_ = this->get_parameter("bitrate").as_int();

        rclcpp::QoS qos_profile(1);
        qos_profile.best_effort();
        qos_profile.keep_last(1);

        // Subscreve ao tópico que o nó de captura está a publicar
        subscription_ = this->create_subscription<sensor_msgs::msg::Image>(
            "/camera/real/image_raw",
            qos_profile,
            std::bind(&VideoEncoderTX::image_callback, this, std::placeholders::_1)
        );

        RCLCPP_INFO(this->get_logger(), "Encoder iniciado. A aguardar imagens... (Destino: %s:%d)", 
                    ip_address_.c_str(), port_);
    }

    ~VideoEncoderTX() {
        if (writer_.isOpened()) {
            writer_.release();
        }
    }

private:
    std::string ip_address_;
    int port_;
    int bitrate_;

    rclcpp::Subscription<sensor_msgs::msg::Image>::SharedPtr subscription_;
    cv::VideoWriter writer_;
    bool writer_initialized_;

    void image_callback(const sensor_msgs::msg::Image::SharedPtr msg) {
        try {
            cv_bridge::CvImagePtr cv_ptr = cv_bridge::toCvCopy(msg, sensor_msgs::image_encodings::BGR8);
            cv::Mat frame = cv_ptr->image;

            if (frame.empty()) return;

            if (!writer_initialized_) {
                int width = frame.cols;
                int height = frame.rows;
                int fps = 30;
                
                std::string pipeline = 
                    "appsrc is-live=true do-timestamp=true ! "
                    "videoconvert ! "
                    "queue ! "
                    "video/x-raw,format=I420 ! "
                    "x264enc tune=zerolatency speed-preset=superfast "
                    "sliced-threads=true threads=1 "
                    "key-int-max=15 intra-refresh=true "
                    "bitrate=" + std::to_string(bitrate_) + " ! "
                    "h264parse config-interval=-1 ! "
                    "rtph264pay pt=96 mtu=1400 aggregate-mode=zero-latency ! "
                    "udpsink host=" + ip_address_ + " port=" + std::to_string(port_) + " sync=false";
                    
                writer_.open(pipeline, cv::CAP_GSTREAMER, 0, fps, cv::Size(width, height), true);
                
                if (!writer_.isOpened()) {
                    RCLCPP_ERROR(this->get_logger(), "Falha ao abrir GStreamer");
                    return;
                }
                writer_initialized_ = true;
            }

            writer_.write(frame);
            
        } catch (cv_bridge::Exception& e) {
            RCLCPP_ERROR(this->get_logger(), "Erro cv_bridge: %s", e.what());
        }
    }
};

int main(int argc, char * argv[]) {
    rclcpp::init(argc, argv);
    rclcpp::spin(std::make_shared<VideoEncoderTX>());
    rclcpp::shutdown();
    return 0;
}